### DISCUSSION TITLE
RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn

### DISCUSSION THREAD

**Discussion by leosongwei** [zai-org/ChatGLM3#834] [Q&A]
运行`finetune_demo/finetune_hf.py`时遇到了这个`RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn`报错。（commit hash 9c875fbd673cc6b9ded5fedff22a5629a86177b3）

查到了这两个issue：

* https://github.com/huggingface/transformers/issues/21381
* https://github.com/huggingface/transformers/pull/27020

有人遇到这个情况的吗？

我的workaround是注释掉503行的`model.gradient_checkpointing_enable()` 。感觉不太对？不过的确finetune成功了。

**Comment by Artistzq** [comment]
我也遇到了这个情况。请问你finetune后的效果怎么样？

**Reply by leosongwei** [reply]
我这里finetune效果出乎意料地好

**Comment by baifachuan** [comment]
我也遇到一样的问题。。这样修改后，结果是不对的，完全finetune不出来，所以应该不是这个改法，但是问题的原因还不名。。

**Comment by youbingchenyoubing** [comment]
一样的问题: RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
![image](https://github.com/THUDM/ChatGLM3/assets/10653481/3702f102-af78-4026-8512-ee7f007e8a84)
![image](https://github.com/THUDM/ChatGLM3/assets/10653481/1b542090-a258-4eb1-aa87-38a7a2471e91)

**Comment by InfiniteInterest** [comment]
我也遇到了，有人解决了么？怎么处理的啊?

**Reply by InfiniteInterest** [reply]
这个问题，我把模型文件全部更新到最新的版本,使用lora微调执行成功了。但是使用ptuning_v2.yaml方式报错, 有没大佬帮看看啊
│                                                                                                  │
│ /home/liushuai/.conda/envs/chatglm3_env/lib/python3.10/site-packages/rouge_chinese/rouge.py:129  │
│ in _get_scores                                                                                   │
│                                                                                                  │
│   126 │   │   │                                                                                  │
│   127 │   │   │   for m in self.metrics:                                                         │
│   128 │   │   │   │   fn = Rouge.AVAILABLE_METRICS[m]                                            │
│ ❱ 129 │   │   │   │   sc = fn(                                                                   │
│   130 │   │   │   │   │   hyp,                                                                   │
│   131 │   │   │   │   │   ref,                                                                   │
│   132 │   │   │   │   │   raw_results=self.raw_results,                                          │
│                                                                                                  │
│ /home/liushuai/.conda/envs/chatglm3_env/lib/python3.10/site-packages/rouge_chinese/rouge.py:54   │
│ in <lambda>                                                                                      │
│                                                                                                  │
│    51 class Rouge:                                                                               │
│    52 │   DEFAULT_METRICS = ["rouge-1", "rouge-2", "rouge-l"]                                    │
│    53 │   AVAILABLE_METRICS = {                                                                  │
│ ❱  54 │   │   "rouge-1": lambda hyp, ref, **k: rouge_score.rouge_n(hyp, ref, 1, **k),            │
│    55 │   │   "rouge-2": lambda hyp, ref, **k: rouge_score.rouge_n(hyp, ref, 2, **k),            │
│    56 │   │   "rouge-3": lambda hyp, ref, **k: rouge_score.rouge_n(hyp, ref, 3, **k),            │
│    57 │   │   "rouge-4": lambda hyp, ref, **k: rouge_score.rouge_n(hyp, ref, 4, **k),            │
│                                                                                                  │
│ /home/liushuai/.conda/envs/chatglm3_env/lib/python3.10/site-packages/rouge_chinese/rouge_score.p │
│ y:253 in rouge_n                                                                                 │
│                                                                                                  │
│   250 │     ValueError: raises exception if a param has len <= 0                                 │
│   251 │   """                                                                                    │
│   252 │   if len(evaluated_sentences) <= 0:                                                      │
│ ❱ 253 │   │   raise ValueError("Hypothesis is empty.")                                           │
│   254 │   if len(reference_sentences) <= 0:                                                      │
│   255 │   │   raise ValueError("Reference is empty.")                                            │
│   256                                                                                            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ValueError: Hypothesis is empty.

**Comment by misslxs** [comment]
一样遇到了

**Comment by youbingchenyoubing** [comment]
model.gradient_checkpointing_enable()的作用
这个函数调用启用了模型的梯度检查点。梯度检查点是一种优化技术，可用于减少训练时的内存消耗。通常，在反向传播期间，模型的中间激活值需要被保留以计算梯度。启用梯度检查点后，系统只需在需要时计算和保留一部分中间激活值，从而减少内存需求。这对于处理大型模型或限制内存的环境中的训练任务非常有用。从它的作用来看, 去掉应该是没影响,至于为什么加上这个不行,我设置了一些断点进行比较.发现是导致loss的梯度计算没有了
![image](https://github.com/THUDM/ChatGLM3/assets/10653481/ae16a52c-c338-402a-80ca-b666e3ae02c0)
![image](https://github.com/THUDM/ChatGLM3/assets/10653481/6ae9aba8-b85b-4e27-999e-b3c25f162e4f)

**Comment by longmans** [comment]
try this:
https://github.com/THUDM/ChatGLM3/pull/846
